### ISSUE TITLE
Feature request: Websocket Support

### ISSUE THREAD

**Issue by 4F2E4A2E** [foss42/api-dash#15]
I would love to have Websocket support.
Is this on your roadmap?

**Comment by ashitaprasad** [comment]
Hi @4F2E4A2E,
Thank you for raising this feature request. We do have Websocket support in our roadmap.

**Comment by mmjsmohit** [comment]
**WebSocket Implementation Request Model:**

The request model for implementing WebSocket should include the following parameters:

- **url**
- **name**
- **description**
- **requestTabIndex**
- **requestHeaders**
- **requestParams**
- **requestBody**
- **requestBodyContentType**

Additionally, each request model should have a corresponding **responseModel** to capture the received response from the sent requests.

**Response Handling:**

We can maintain a list of such requests, which can be displayed in the response pane alongside the corresponding response. Each response should consist of:

- **message**
- **status**

**Interface Sketch:**

For a visual representation, please refer to the rough sketch of the interface. I have also included some more details in the sketch: [Sketch](https://excalidraw.com/#json=0glbl5eMbsBBil00PrkGs,qJGK11FkPZaBW0cTiGLUug)

**Comment by ashitaprasad** [comment]
Hi @mmjsmohit,
Thanks for taking up this issue and posting and initial draft of implementation details.
Did you get a chance to take a look at other GUI API clients to understand the different features people use to test WebSocket?

**Comment by mmjsmohit** [comment]
Yes, @ashitaprasad! I have looked at Postman and Thunder Client. Are there any other tools that I should check out?

**Comment by mmjsmohit** [comment]
@ashitaprasad @animator Should I go ahead with the implementation outlined in the above sketch?

**Comment by animator** [comment]
Yes @mmjsmohit You can go ahead and draft an initial implementation as shown in the sketch you provided.
![Untitled-2024-02-24-2211-2](https://github.com/foss42/apidash/assets/615622/acad96ed-2426-427b-8e1b-e1db170bac35)

**Comment by mmjsmohit** [comment]
@animator @ashitaprasad I made respectable progress in solving the issue. During the process, I realized that the response models for WebSocket and HTTP requests differ. Should I create new models for each protocol?

**Comment by ashitaprasad** [comment]
Yes @mmjsmohit the request and response models for WebSocket will be different.
You need to create a new folder inside models for websocket and keep the relevant models inside it.

**Comment by mmjsmohit** [comment]
> > @animator @ashitaprasad I made respectable progress in solving the issue. During the process, I realized that the response models for WebSocket and HTTP requests differ. Should I create new models for each protocol?
> 
> Sorry @mmjsmohit, I did not see your message as I was already playing with the implementation. I opened up a PR with a working example and some explanation.

I have also already implemented the whole feature and updates @animator and @ashitaprasad on Discord about the same. 

Please let me know if I should open a PR or not.

**Comment by ashitaprasad** [comment]
@mmjsmohit, you can go ahead and raise a PR.

**Comment by Clasherzz** [comment]
@ashitaprasad I would like to contribute on this and have some previous experiences working on websocket projects and understand the codebase well enough to make an approach. Is the feauture currently under development and is free to work on?

**Comment by ashitaprasad** [comment]
@Clasherzz You are free to work on this feature .. it is recommended that you share the feature designs first.

**Comment by Clasherzz** [comment]
![image](https://github.com/user-attachments/assets/94a8c08d-927e-457f-9f6a-cc838d12266d)

**Comment by Clasherzz** [comment]
@ashitaprasad  looking forward to implement some cards for each messages . Each card could be expaned to see the contents . would have incoming / outtgoing arrow. Messages would have unique ids based on their time stamp  which can be used for deletion.
The params and headers are disabled after the initial hanshake . The message would have dropdown supporting text and binary .

Looking forward to use https://fluttergems.dev/packages/web_socket_channel/  . This got all the functionalities we needed and also is the most used one for our need.

Now would it possinle to know whether this feauture is being implemented from the organizations side and whether its almost being done completed. I would like to know of this because the earlier PR i worked on graphql was closed . I would like to participate in next gsoc and I would like to submit a good quality PR essential to the organization to improve my profile. But if the feauture is closed at the last moment then I would be left with no PRs to show my work . Hope you understand.

**Comment by ashitaprasad** [comment]
Hi @Clasherzz, The PR with the best solution is always accepted into the project whether it comes from the maintainers or other contributors. Also, we follow certain guidelines while accepting PRs like code maintainability, whether it is scalable, whether the code submitted is clean, amount of code changes done and how well tested the code is. We provided you feedback on your PR along these aspects and we recommend you to learn from this feedback. Since, it is an open source project we cannot guarantee whether other contributors will not work on the same issue.

**Comment by Clasherzz** [comment]
@ashitaprasad  no i didnt mean to sound prude. I am fully aware that many people can work on the same issue at the time and the best one would be choosen. The thing with the last PR was that it was a direct PR from you . So i thought i can ask you directly if this being worked on by your side which i hope is not a wrong question to ask. 

@ashitaprasad also i would love to know if the implementation  i mentioned above is something i can work on.

**Comment by Clasherzz** [comment]
New layout 

![Image](https://github.com/user-attachments/assets/185911e7-77ea-4b53-8015-7efc2e2f90c7)

**Comment by AnujaKalahara99** [comment]
Hi Team, I recently started to try out API Dash and did not found any async API testing tools? Isn't this feature in the main build yet?

**Comment by ShashwatXD** [comment]
@animator @ashitaprasad , picking this up!

**Comment by ShashwatXD** [comment]
- Added full WebSocket support with ws:// and wss:// request handling.
- Implemented connect / disconnect lifecycle with real-time message exchange.
- Introduced dedicated WebSocket Request pane and chat-style Response pane with timestamps.
- Added connection state toggle logic to prevent invalid or stale connections.

Checkout this video below: (used `wss://echo.websocket.org/` for verification)

https://github.com/user-attachments/assets/af8277ee-f8b0-4a6c-a297-27d90b8965e4

Please let me know what improvements we might need, rest i will raise a pr once this looks fine @animator 

Edits/Updates:
- I have added failed state handling
- added history preservation for websockets

**Comment by Rehyanyadav** [comment]
Hi @ashitaprasad @animator, I would like to work on WebSocket support

(Issue #15) for GSOC 2026. I've done extensive research on the existing

PRs (#1003, #1017), the GSOC 2025 proposals, and the architecture.

Below is my detailed 12-week implementation proposal. Would appreciate

your feedback on the approach and scope before I begin coding.


## Overview
I am proposing to take on **WebSocket support (Issue #15)** as my GSOC 2026 contribution to API Dash. This is a high-priority L1 roadmap item that will bring real-time protocol testing capabilities to the platform.

## Current State Analysis
After reviewing the existing work:
- **PR #1003** (ShashwatXD): Full WebSocket support with ws:// and wss:// handling, connection lifecycle, message exchange, and chat-style UI
- **PR #1017** (Kan7sh): Another WebSocket implementation
- Multiple past PRs and proposals showing strong community interest
- **GSOC 2025 proposals** (Kushagra, K_Govind, RafsanNeloy) outline detailed WebSocket + SSE + MQTT + gRPC roadmaps

### Phase 1: Foundation & Code Review (Weeks 1-2)
- Review existing PRs (#1003, #1017) and analyze feedback from maintainers
- Understand current architecture: request models, response handling, Riverpod providers, code generation
- Set up local dev environment and run all integration tests
- Study `web_socket_channel` (^3.0.3) and dart:io WebSocket for cross-platform support

### Phase 2: Core WebSocket Implementation (Weeks 3-4)
If starting fresh or refining existing:
- **WebSocket Handler**: Abstract interface + concrete implementation
  - Connection establishment (ws://, wss://)
  - Send/receive message flow
  - Connection state management (connecting, connected, disconnected, error)
  - Automatic reconnection with configurable intervals & attempts
  
- **Request/Response Models**:
  - `WebSocketRequest`: URL, headers, subprotocols, body, name, description
  - `WebSocketResponse`: message log with timestamps, status, sender (client/server)
  - History persistence per request tab

- **Riverpod State Management**:
  - `websocketHandlerProvider` - manage connection lifecycle
  - `websocketMessagesProvider` - stream incoming messages
  - `websocketConnectionStatusProvider` - connection state UI binding

### Phase 3: UI Implementation (Weeks 5-6)
- **WebSocket Request Pane**: URL input, headers, body, connect/disconnect toggle
- **Response Pane**: Chat-style message display with timestamps, sender badges
- **Controls**: Search messages, clear all/individual messages, scroll to top/bottom
- **Settings Integration**: Ping interval, reconnection attempts, retry intervals
- **Cross-platform**: Ensure desktop/mobile responsiveness

### Phase 4: Code Generation (Weeks 7-8)
- Extend `codegen/` templates for WebSocket in supported languages (Python, JavaScript, Java, Go, etc.)
- Generate example code for connecting, sending/receiving messages
- Example: Python with `websocket-client`, JS with native WebSocket, etc.

### Phase 5: Testing & Refinement (Weeks 9-10)
- **Unit Tests**: WebSocket handler logic, message parsing, connection state transitions
- **Integration Tests**: Widget tests for UI, real flows with public echo servers (wss://echo.websocket.org)
- **Cross-platform Testing**: Validate on macOS, Windows, Linux, Android, iOS
- **Performance**: Monitor CPU/memory with high-volume message streams
- **Error Handling**: Connection failures, invalid URLs, timeouts, disconnects

### Phase 6: Documentation & Polish (Weeks 11-12)
- Update [User Guide](doc/user_guide/README.md) with WebSocket testing examples
- Add developer docs explaining architecture decisions
- Create brief tutorial video or screenshots
- Address code review feedback, finalize PR for merge

## Why This Topic?

 **High Impact**: WebSocket is L1 priority, needed for real-time API testing (chat, live feeds, IoT)  
 **Clear Scope**: Well-defined protocol, existing PRs show path forward  
 **Full-Stack**: HTTP → WebSocket protocol, Riverpod state, Flutter UI, code gen, tests  
 **Community Demand**: 3+ past GSOC proposals, 2+ active PRs  
 **Sustainable**: Positions API Dash for SSE, MQTT, gRPC expansions next  

## Technical Stack

**WebSocket**: Uses web_socket_channel for reliable, cross-platform support
**State Management**: Riverpod for managing connection state and message streams
**UI**: Built with Flutter for desktop and mobile compatibility
**Testing**: Unit and UI tests using Flutter and Riverpod testing tools
**Code Generation**: Extends existing templates to stay consistent with REST and GraphQL APIs


## Success Criteria

-  WebSocket request/response handling works end-to-end
-  UI is intuitive and responsive on desktop/mobile
-  Code generation produces working code in ≥3 languages
-  Unit + integration tests cover main flows (80%+ coverage)
-  PR passes all maintainer reviews and is merged
-  Documentation is clear and examples are runnable

**Comment by Wreck-X** [comment]
@animator Hi is this being worked on currently?, id like to take a look at it.

**Comment by Rehyanyadav** [comment]
yes i'm working on it

**Comment by Rehyanyadav** [comment]
@animator i have complete the coding and testing part  can you review and check the web socket support

**Comment by roshan-melvin** [comment]
Hi @animator sir and @ashitaprasad mam,

Following up on the feedback and issue placed in [#1529](https://github.com/foss42/apidash/issues/1529) , here is the video demonstration showcasing the native WebSocket protocol implementation in API Dash for this issue:


https://github.com/user-attachments/assets/3ace0f45-dfa4-4cf6-86e5-02370168757b

I have also opened a new, updated PR ([#1613](https://github.com/foss42/apidash/pull/1613)) which includes this implementation along with the video demos in the description, as requested. Please let me know if there are any specific connection scenarios or edge cases you would like me to demonstrate!

Thanks!
